crates/proof_of_stake/src/lib.rs

```diff
@@ -2913,3 +2948,110 @@ where
 
     Ok(())
 }
+
+fn add_delegation_target<S>(
+    storage: &mut S,
+    delegator: &Address,
+    validator: &Address,
+    epoch: Epoch,
+    _current_epoch: Epoch,
+) -> namada_storage::Result<()>
+where
+    S: StorageRead + StorageWrite,
+{
+    let bond_holders = delegation_targets_handle(delegator);
+    if let Some(delegations) = bond_holders.get(storage, validator)?.as_mut() {
+        let (start, end) = delegations.last_range;
+        if let Some(end) = end {
+            // Add the `last_range` pair to the `prev_ranges` and make a new
+            // `last_range`
+            if epoch == end {
+                // This case would occur if in the same epoch, the bond was
+                // fully unbonded, followed by the bonding of new tokens
+                delegations.last_range.1 = None;
+            } else {
+                delegations.prev_ranges.insert(start, end);
+                delegations.last_range = (epoch, None);
+            }
+            bond_holders.insert(
+                storage,
+                validator.clone(),
+                delegations.clone(),
+            )?;
+        } else {
+            // do nothing since the last bond is still active
+        }
+    } else {
+        // Make a new delegation to this source-validator pair
+        let first_delegation = DelegationEpochs {
+            prev_ranges: BTreeMap::new(),
+            last_range: (epoch, None),
+        };
+        bond_holders.insert(storage, validator.clone(), first_delegation)?;
+    }
+
+    // Todo: possibly update the data by pruning old pairs in
+    // `delegations.prev_ranges`??
```

Comment:
should we call prune here?

Reply:
I was thinking of maybe only doing it in `remove_delegation_target` just to keep this lean? After all, `prev_ranges` will only grow the more one calls `remove_delegation_target`. What do you think?

Reply:
Ah I see, yeah sounds good!

Reply:
Could you put this reasoning into a comment here pls